BitcoinJ custom logic continuation

stanbar opened the pull request:
wallet encryption and decryption by key derivation
remove Password class
new webapp design
closes #75 
WebSocket communication with conflated channels thanks to: https://github.com/ktorio/ktor/issues/717
autolock also clears module inputs
move common dependency to parent pom
add CommunicationChannel via kotlin coroutines channels so backend -> frontend communication is possible with ease
fixed bug where prepareShutdownHook(bitcoin); was never called because startKtorServer is blocking function.
backend communicate error, info, success messages to frontend via CommunicationChannel

Took 4 hours 20 minutes

stanbar committed:
add websockets

Took 18 minutes

stanbar committed:
bidirectional communication via Websocket
Send blockchain download progress to WebApp over ws

Took 57 minutes

stanbar committed:
wallet encryption and decryption by key derivation
remove Password class
new webapp design

Took 2 hours 3 minutes

PatrykMilewski commented:
I will review it tommorow at 11-12 am, remind me if I don't and it will block other things

stanbar committed:
WebSocket communication with conflated channels thanks to: https://github.com/ktorio/ktor/issues/717

Took 33 minutes

stanbar committed:
autolock also clears module inputs

Took 5 minutes

stanbar committed:
refactor maven multi-modules
move common dependency to parent pom
encapsulate AutoLockTimer
add CommunicationChannel via kotlin coroutines channels so backend -> frontend communication is possible with ease
fixed bug where prepareShutdownHook(bitcoin); was never called because startKtorServer is blocking function.
backend communicate error, info, success messages to frontend via CommunicationChannel

Took 1 hour 44 minutes

stanbar committed:
implement @cy6erGn0m suggestions

Took 16 minutes

PatrykMilewski commented on Manager/src/main/java/io/raspberrywallet/manager/Manager.java:
```diff
@@ -49,34 +47,39 @@
+            try {
+                lockWallet();
+            } catch (WalletNotInitialized ignored) {
+                //we don't care about locking if it wasn't even inited
```
cannot these two try catch blocks merge into one? I can't find any reason to split them

stanbar replied:
They can not be merged because if the first line throw exception then the `autoLock()` won't be executed. Anyway, you realized me that here is a different bug.  The `frontendChannel.info();` doesn't even throws any exception 😆 Fixed.

PatrykMilewski commented on Manager/src/main/java/io/raspberrywallet/manager/bitcoin/Bitcoin.java:
```diff
@@ -177,6 +186,15 @@ private void synchronizeWalletNonBlocking(final Wallet wallet, @Nullable KeyPara
+                if (password != null)
+                    saveEncryptedWallet(password);
+                Logger.info("Wallet balance" + wallet.getBalance().toFriendlyString());
+            } catch (WalletNotInitialized | IOException e) {
```
this method throws IOException anyways, maybe you can just throw it, instead of catching and printing stack trace

stanbar replied:
Unfortunately, Java doesn't allow Lambdas to throw exceptions, it must be caught inside it.
A workaround is to use Kotlin ❤️ 🤣

PatrykMilewski commented on Manager/src/main/java/io/raspberrywallet/manager/bitcoin/WalletCrypter.java:
```diff
@@ -0,0 +1,55 @@
+            wallet.decrypt(aesKey);
+        } else {
+            Logger.err("User entered incorrect password");
+            Logger.err("Wrong password");
```
these two log lines are actually the same

stanbar replied:
Right, fixed
